### PR TITLE
Issue #3775: [Performance] Optimize node access query building

### LINKED ISSUE #3775: [Performance] Optimize node access query building
This is the equivalent for https://www.drupal.org/project/drupal/issues/106721 (D7 core), it is currently RTBC, and is blocking https://www.drupal.org/node/2204257 (Views 7.x). I came across it while working on the Views crossports in #3720.

> **Problem/Motivation**
> Drupal's current method of determining whether a specified user has access to a node is reported to have significant performance issues on a site using node access with a large number of grants.
> 
> **Proposed resolution**
> A patch has been prepared, however it appears to be unclear as to what functionality/performance testing might be required or how it would be undertaken.
> 
> **Remaining tasks**
> (todo) commit patch [#133](https://www.drupal.org/project/drupal/issues/106721#comment-10750082)

### PR DESCRIPTION
**Pull request by klonos**
Fixes https://github.com/backdrop/backdrop-issues/issues/3775

------

https://www.drupal.org/project/drupal/issues/106721

https://www.drupal.org/files/issues/drupal-106721-optimize_node_access_queries-133.patch (RTBC)

### CODE REVIEW

**Comment by klonos** [comment]
...I have included https://git.drupalcode.org/project/views/commit/d649d329eb2fc56d9feb1320c173fd69f3c9a986, which is part of https://github.com/backdrop/backdrop-issues/issues/3720, but requires the newly-introduced `node_add_node_grants_to_query()` function from https://www.drupal.org/project/drupal/issues/106721

**Comment by jenlampton** [comment]
This change hasn't been committed to views yet. (well, it was, but then it was reverted)

And https://www.drupal.org/project/drupal/issues/106721 hasn't been committed to core yet. I think we should wait on this until those go in.

**Comment by quicksketch** [comment]
> This change hasn't been committed to views yet. (well, it was, but then it was reverted)
> 
> And https://www.drupal.org/project/drupal/issues/106721 hasn't been committed to core yet. I think we should wait on this until those go in.

Looks like the Drupal core commit has been in for years now. The [Views issue ](https://www.drupal.org/node/2204257) is however unlikely to ever be merged at this point. It looks like it was reverted because Drupal core needed to make the update first, and it was too early to make the change in Views. With Views being a part of Core in Backdrop, that won't be a problem for us.

Looks like a simple and nice set of changes here, and with it in production on D7 for a while now we should have a high degree of confidence.

File: core/modules/node/node.module [hunk 1]
```diff
@@ -2766,6 +2766,31 @@ function node_access($op, $node, $account = NULL) {
...
 
+/**
+ * Helper function to create the OR condition for a node grants check.
+ *
+ * @param $account
+ *   The grants to add to the query, usually gotten via node_access_grants().
+ * @param $table_alias
+ *   Optional, the alias to the node access table.
+ *
+ * @return
```
Comment (quicksketch) [inline core/modules/node/node.module]: Looks like the parameter documentation is mismatched here.

```suggestion
 * @param array $node_access_grants
 *   The grants to add to the query, usually gotten via node_access_grants().
 * @param string $table_alias
 *   Optional, the alias to the node access table.
 *
 * @return bool
```

File: core/modules/node/node.module [hunk 2]
```diff
@@ -2681,6 +2681,31 @@ function node_access($op, $node, $account = NULL) {
...
+ * Helper function to create the OR condition for a node grants check.
+ *
+ * @param array $node_access_grants
+ *   The grants to add to the query, usually gotten via node_access_grants().
+ * @param string $table_alias
+ *   Optional, the alias to the node access table.
+ *
+ * @return bool
+ *   TRUE if the operation may be performed, FALSE otherwise.
+ */
```
Comment (quicksketch) [inline core/modules/node/node.module:2696]:
```suggestion
 *
 * @since 1.30.0 Function added.
 */
```

**Comment by quicksketch** [comment]
Thanks folks! I merged this into 1.x for 1.30.0. I added a `@since` docblock line for 1.30.0.

**Pull request merged**